Utilities/ocvalidate/ocvalidate.c

```diff
@@ -34,41 +37,1272 @@
-    printf("Read fail\n");
-    return -1;
+STATIC
+CHAR8 *
+GetFilenameSuffix (
+  IN  CONST CHAR8  *FileName
+  )
+{
+  CONST CHAR8  *SuffixDot;
+
+  //
+  // Find the last dot by reverse order.
+  //
+  SuffixDot = OcAsciiStrrChr (FileName, '.');
+
+  //
+  // In some weird cases the filename can be crazily like
+  // "a.suffix." (note that '.') or "..." (extremely abnormal).
+  // We do not support such.
+  //
+  if (!SuffixDot || SuffixDot == FileName) {
+    return (CHAR8 *) "";
+  }
+
+  //
+  // In most normal cases, return whatever following the dot.
+  //
+  return (CHAR8 *) SuffixDot + 1;
+}
+
+STATIC
+BOOLEAN
+AsciiStringHasAllLegalCharacter (
+  IN  CONST CHAR8  *String
+  )
+{
+  UINTN  Index;
+
+  for (Index = 0; Index < AsciiStrLen (String); ++Index) {
+    //
+    // Skip allowed characters (0-9, A-Z, a-z, '_', '-', '.', '/', and '\').
+    //
+    if (IsAsciiNumber (String[Index])
+      || IsAsciiAlpha (String[Index])
+      || String[Index] == '_'
+      || String[Index] == '-'
+      || String[Index] == '.'
+      || String[Index] == '/'
+      || String[Index] == '\\') {
+      continue;
+    }
+
+    //
+    // Disallowed characters matched.
+    //
+    return FALSE;
   }
 
-  long long a = current_timestamp();
+  return TRUE;
+}
+
+STATIC
+BOOLEAN
+AsciiStringHasAllPrintableCharacter (
+  IN  CONST CHAR8  *String
+  )
+{
+  UINTN  Index;
+
+  for (Index = 0; Index < AsciiStrLen (String); ++Index) {
+    //
+    // Unprintable characters matched.
+    //
+    if (IsAsciiPrint (String[Index]) == 0) {
+      return FALSE;
+    }
+  }
+
+  return TRUE;
+}
+
+STATIC
+UINT32
+ReportError (
+  IN  CONST CHAR8  *FuncName,
+  IN  UINT32       ErrorCount
+  )
+{
+  if (ErrorCount != 0) {
+    DEBUG ((DEBUG_WARN, "%a returns %u %a!\n", FuncName, ErrorCount, ErrorCount > 1 ? "errors" : "error"));
+  } else {
+    DEBUG ((DEBUG_VERBOSE, "%a returns no errors!\n", FuncName));
+  }
+
+  return ErrorCount;
+}
+
+STATIC
+BOOLEAN
+DataHasProperMasking (
+  IN  CONST VOID   *Data,
+  IN  CONST VOID   *Mask,
+  IN        UINTN  Size
+  )
+{
+  CONST UINT8  *ByteData;
+  CONST UINT8  *ByteMask;
+  UINTN        Index;
+
+  ByteData = Data;
+  ByteMask = Mask;
+
+  for (Index = 0; Index < Size; ++Index) {
+    if ((ByteData[Index] & ~ByteMask[Index]) != 0) {
+      return FALSE;
+    }
+  }
+
+  return TRUE;
+}
+
+STATIC
+UINT32
+CheckACPI (
+  IN  OC_GLOBAL_CONFIG  *Config
+  )
+{
+  UINT32          ErrorCount;
+  UINT32          Index;
+  OC_ACPI_CONFIG  UserAcpi;
+  CONST CHAR8     *Path;
+  CONST CHAR8     *Comment;
+  CONST UINT8     *Find;
+  UINT32          FindSize;
+  CONST UINT8     *Replace;
+  UINT32          ReplaceSize;
+  CONST UINT8     *Mask;
+  UINT32          MaskSize;
+  CONST UINT8     *ReplaceMask;
+  UINT32          ReplaceMaskSize;
+  BOOLEAN         HasCustomDSDT;
+
+  DEBUG ((DEBUG_VERBOSE, "config loaded into ACPI checker!\n"));
+
+  ErrorCount    = 0;
+  UserAcpi      = Config->Acpi;
```

Comment:
Please do not copy large structures by assignment. Simply make it a pointer.

Reply:
Done. https://github.com/acidanthera/OpenCorePkg/pull/166/commits/0f8374903c91848200cacfb4a531c509ecaca94e